archive/writer.go

```diff
@@ -0,0 +1,50 @@
+package archive
+
+import (
+	"archive/tar"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type TarWriter interface {
+	WriteHeader(hdr *tar.Header) error
+	Write(b []byte) (int, error)
+	Close() error
+}
+
+type NormalizingTarWriter struct {
+	TarWriter
+	headerOpts []HeaderOpt
+}
+
+type HeaderOpt func(header *tar.Header) *tar.Header
+
+func (tw *NormalizingTarWriter) WithUID(uid int) {
+	tw.headerOpts = append(tw.headerOpts, func(hdr *tar.Header) *tar.Header {
+		hdr.Uid = uid
+		return hdr
+	})
+}
+
+func (tw *NormalizingTarWriter) WithGID(gid int) {
+	tw.headerOpts = append(tw.headerOpts, func(hdr *tar.Header) *tar.Header {
+		hdr.Gid = gid
+		return hdr
+	})
```

Comment:
I wonder if eventually we might want to unify (or move and duplicate) this related imgutil [WindowsWriter logic](https://github.com/buildpacks/imgutil/blob/2281cd9b141467576e74acc76d80bddf276189de/layer/windows_writer.go#L54-L59) to these `WithGID` and `WithUID` methods. 

Definitely doesn't need to happen right now because the `WindowsWriter` just looks at the `hdr.Gid/Uid` and applies generic owners/groups instead of specific users via PAXRecords. But for feature parity, we may eventually need a more robust implementation where `WithUID` and `WithGID` can take specific SIDs and apply them Windows Owner and Primary group permissions. 

Nothing to do now, just something we can keep in mind.

Reply:
Agree, that makes a lot of sense. In the lifecycle we are always creating windows layers while running in a windows environment. That is not true for `pack` so we will have to be thoughtful about that when we move them over as a generic utility.